bazarr/subtitles/refiners/__init__.py

```diff
@@ -4,10 +4,12 @@
 from .database import refine_from_db
 from .arr_history import refine_from_arr_history
 from .anidb import refine_from_anidb
+from .anilist import refine_from_anilist
 
 registered = {
     "database": refine_from_db,
     "ffprobe": refine_from_ffprobe,
     "arr_history": refine_from_arr_history,
     "anidb": refine_from_anidb,
+    "anilist": refine_from_anilist, # Must run AFTER AniDB
```

Comment:
Im curious why does this needs to run after anidb?

Reply:
The refiner piggybacks off the AniDB refiner, utilizing the attribute [`series_anidb_id`](https://github.com/morpheus65535/bazarr/blob/8dd7ab4ae41d3f7c10115bea908d6f20a36705ef/bazarr/subtitles/refiners/anilist.py#L70) of `video` which gets populated using said refiner.  
If `anilist` is run before `anidb`, this attribute won't have a value and thus the refiner won't work with a show that has a higher season than 1.

Reply:
I think essentially i find weird why the anilist depends of anidb since they are completely different stuff.

Also why it doesn't work for season higher than 1? anilist should have them all

Reply:
I see what you are doing, you are using the mapping of the anidb, that has season to find the anidb_id and go to the other mapping using this anidb_id 🤔

Reply:
anidb refiner depends on the api, what you could do to remove the dependency would be to also download the same mapping and always use the anidb_id then

Reply:
Yes, both are different things, but your refiner adds crucial information to the `video` on which I depend on.

For instance, if I were to get the AniList ID for season 3 of "Tensei Shitara Suraimu..." using only the TVDB ID, I'd get a bunch of entries back with the [mapping list](https://github.com/Fribb/anime-lists) I use:
```PowerShell
❯ $list | ? thetvdb_id -eq 352408 | select anime-planet_id, thetvdb_id, anilist_id

anime-planet_id                                            thetvdb_id anilist_id
---------------                                            ---------- ----------
that-time-i-got-reincarnated-as-a-slime                        352408     101280
that-time-i-got-reincarnated-as-a-slime-season-2               352408     108511
that-time-i-got-reincarnated-as-a-slime-season-2-part-ii       352408     116742 # <- I want this
that-time-i-got-reincarnated-as-a-slime-season-3               352408     156822
that-time-i-got-reincarnated-as-a-slime-movie-scarlet-bond     352408     139498
that-time-i-got-reincarnated-as-a-slime-ova                    352408     106509
that-time-i-got-reincarnated-as-a-slime-visions-of-coleus      352408     161802
```

However, your refiner adds an AniDB ID which is useful for my list.  
Having this information allows me to derive the correct AL ID:  
```PowerShell
❯ $a | ? anidb_id -eq 17709 | select anime-planet_id, anilist_id, anidb_id

anime-planet_id                                  anilist_id anidb_id
---------------                                  ---------- --------
that-time-i-got-reincarnated-as-a-slime-season-3 156822    17709
```

At first I also wanted to work with the `Anime-Lists/anime-lists` XML, but then I noticed that [you've been using that list](https://github.com/morpheus65535/bazarr/blob/008e05b71be22ea6d79d8d45b48cd472ae13578d/bazarr/subtitles/refiners/anidb.py#L35) yourself already.  
To avoid redundant code, I opted to piggyback off your refiner, knowing that it would resolve an AniDB ID first.  
Ultimately, that's the reason why I depend on the AniDB refiner.

Reply:
Yeah okay, i'm fine as it is, i don't have a better idea tho.

Reply:
If we could extract the logic of anidb_seriesid for both refines to use would be the ideal.

Reply:
I guess we can refactor the anidb refiner to don't prevent adding the series_id even if the api key is not set

Reply:
> I guess we can refactor the anidb refiner to don't prevent adding the series_id even if the integration is not set, and validate afterwards or skip.

Sounds good, I'll need to merge with the upstream changes anyway and in the same vein could do this refactoring.  
Also perhaps it's best if this logic doesn't get extracted, as otherwise we'd have multiple places doing AniDB stuff which could become confusing.

Reply:
Refactored in [690ed33](https://github.com/morpheus65535/bazarr/pull/2505/commits/690ed335beefa15e4926552d180a45b45a39a945): `anidb_series_id` will now always get refined, despite no API client having been set up.  
That said, I was unable to fully test the changes with the AnimeTosho provider as their domain is currently down.

Reply:
@anderson-oki We might need to address the AniDB refiner a bit more apart from just refactoring it.  
It seems that the XML we use doesn't always allow us to resolve ambiguity.

Take note of the values for `defaulttvdbseason`:
```PowerShell
❯ $list | ? tvdbid -eq 279827                                      

anidbid tvdbid defaulttvdbseason name
------- ------ ----------------- ----
10445   279827 1                 Aldnoah.Zero
10846   279827 1                 Aldnoah.Zero (2015)

❯ $list | ? tvdbid -eq 352408

anidbid tvdbid defaulttvdbseason name
------- ------ ----------------- ----
13871   352408 1                 Tensei Shitara Slime Datta Ken
14526   352408 0                 Tensei Shitara Slime Datta Ken (2019)
14767   352408 2                 Tensei Shitara Slime Datta Ken (2021)          # <---
15456   352408 2                 Tensei Shitara Slime Datta Ken (2021 Dai 2 Bu) # <---
17709   352408 3                 Tensei Shitara Slime Datta Ken (2024)
```

As you can see, some entries have identical values for the aforementioned key.  
This leads to cases where multiple entries get matched for a given `video.season`, often leading to the wrong `anidbid` getting returned.

I don't have a solution yet; Just leaving this comment here for awareness.

Reply:
Isnt the offset this? I didnt check but i thought the offset would resolve this

Reply:
The example you gave is already covered since my first implementation. Note that they have different offsets.

Reply:
Please test with animetosho as well.

Reply:
I believe to have found the issue, it's due to a [logic error](https://github.com/morpheus65535/bazarr/blob/98c2fb5600976775e17d5654fd21c91c48fbd1c4/bazarr/subtitles/refiners/anidb.py#L92):  
`anidb_id ` always gets overridden in each iteration of `enumerate(animes)`.

Currently working on a commit that fixes this.

Reply:
Yeah you are right, line 86 should be moved to line 88